### PR TITLE
Move fluidapp-odsp-urlresolver

### PR DESCRIPTION
**Pull request by chentong7**
## Description

> This resolver plays a crucial role in converting URLs from OneNote and Office.com for the Fetch tool, which is essential for retrieving snapshots for analysis when needed. Therefore, it's not feasible to eliminate this resolver.

### CODE REVIEW

File: package.json
Comment (alexvy86) [inline package.json:1]: The changes in this file look unrelated, I think they are not intended? Once these are reverted we can look at the lockfile diff to make sure only the expected changes are there.

File: packages/drivers/odsp-urlResolver/package.json [hunk 1]
```diff
@@ -107,7 +109,6 @@
 		"@fluidframework/build-tools": "^0.29.0",
 		"@fluidframework/eslint-config-fluid": "^3.3.0",
 		"@fluidframework/mocha-test-setup": "workspace:~",
-		"@fluidframework/odsp-urlresolver-previous": "npm:@fluidframework/odsp-urlresolver@2.0.0-internal.8.0.0",
```
Comment (alexvy86) [inline packages/drivers/odsp-urlResolver/package.json:110]: Interesting... do we know why this was using a `npm:` dependency instead of a `workspace:` one?

File: packages/drivers/odsp-urlResolver/src/test/types/validateOdspUrlresolverPrevious.generated.ts
Comment (alexvy86) [inline packages/drivers/odsp-urlResolver/src/test/types/validateOdspUrlresolverPrevious.generated.ts:1]: Why did we delete this file? I'd expect we still need type tests for the package that remains, no?

File: feeds/public.txt [hunk 2]
```diff
@@ -44,7 +44,7 @@ fluid-framework
 @fluidframework/odsp-driver-definitions
 @fluidframework/odsp-driver
 @fluidframework/local-driver
-@fluid-tools/fluidapp-odsp-urlresolver
+@fluidframework/odsp-urlresolver
```
Comment (alexvy86) [inline feeds/public.txt:47]: Looks like a manual replacement here? This package is already listed on line 43 (and same for the other two package list files). I think we just need to regenerate these by running a full build from the root, and I'd expect this line just disappears.

**Review by alexvy86 — COMMENTED** [review]
A few comments. Also, can we make the description clearer? I first thought it meant the whole package, but we're still removing it. After reviewing I understood it means the class that was moved.

File: package.json
Comment (chentong7) [reply package.json:1]: I think I messed up with the rebase. Let me re-rebase it

File: packages/drivers/odsp-urlResolver/src/test/types/validateOdspUrlresolverPrevious.generated.ts
Comment (chentong7) [reply packages/drivers/odsp-urlResolver/src/test/types/validateOdspUrlresolverPrevious.generated.ts:1]: The test expecting the previous built package. Deleting it for now and add it back after it has been built to a new previous package

File: packages/drivers/odsp-urlResolver/api-report/odsp-urlresolver.api.md [hunk 3]
```diff
@@ -4,10 +4,19 @@
 
 ```ts
 
+import { IContainerPackageInfo } from '@fluidframework/driver-definitions';
 import { IRequest } from '@fluidframework/core-interfaces';
 import { IResolvedUrl } from '@fluidframework/driver-definitions';
 import { IUrlResolver } from '@fluidframework/driver-definitions';
 
+// @internal (undocumented)
+export class FluidAppOdspUrlResolver implements IUrlResolver {
```
Comment (Josmithr) [inline packages/drivers/odsp-urlResolver/api-report/odsp-urlresolver.api.md:13]: @jatgarg Can you recommend documentation for this? Seems like a good opportunity to fill in some docs while we're moving this.

File: packages/drivers/odsp-urlResolver/package.json (hunk 1 above)
Comment (Josmithr) [reply packages/drivers/odsp-urlResolver/package.json:110]: This is the pattern used for installing "previous" versions of the same package for use in the type-tests.

Is there a reason we're removing it? We still need this in order for the package's type-tests to function.

File: package.json [hunk 4]
```diff
@@ -155,7 +155,7 @@
 	"devDependencies": {
 		"@changesets/cli": "^2.26.1",
 		"@fluid-private/changelog-generator-wrapper": "file:tools/changelog-generator-wrapper",
-		"@fluid-tools/build-cli": "^0.29.0",
```
Comment (jatgarg) [inline package.json:158]: Why any of these changes are here? These does not seem to be related to your change?

File: packages/drivers/odsp-urlResolver/api-report/odsp-urlresolver.api.md (hunk 3 above)
Comment (jatgarg) [reply packages/drivers/odsp-urlResolver/api-report/odsp-urlresolver.api.md:13]: You can put something like, 
This resolver is used to help resolve urls which belongs to onenote, office.com etc.

File: pnpm-lock.yaml
Comment (jatgarg) [inline pnpm-lock.yaml:1]: I am not able to see any changes in pnpm file. Not sure what they are.

File: packages/drivers/odsp-urlResolver/package.json (hunk 1 above)
Comment (chentong7) [reply packages/drivers/odsp-urlResolver/package.json:110]: Seems like those previous packages are generated by type-tests. It got removed somehow along with the package re-built. Running type tests again...

Comment (Josmithr) [reply packages/drivers/odsp-urlResolver/package.json:110]: The previous packages are just the version of the package we specify inline, as published to npm. These dependencies are _used_ by the type-tests, but they aren't created by them.

File: packages/drivers/odsp-urlResolver/api-report/odsp-urlresolver.api.md (hunk 3 above)
Comment (Josmithr) [reply packages/drivers/odsp-urlResolver/api-report/odsp-urlresolver.api.md:13]: Thanks!

Comment (chentong7) [reply packages/drivers/odsp-urlResolver/api-report/odsp-urlresolver.api.md:13]: Updating accordingly

File: package.json (hunk 4 above)
Comment (chentong7) [reply package.json]: Removed it!

File: packages/drivers/odsp-urlResolver/api-report/odsp-urlresolver.api.md (hunk 3 above)
Comment (chentong7) [reply packages/drivers/odsp-urlResolver/api-report/odsp-urlresolver.api.md:13]: I updated the documentation at ts file and rebuild it. The api report's `(undocumented)` mark is gone but I don't see the documentation here. Is that expected?

Comment (chentong7) [reply packages/drivers/odsp-urlResolver/api-report/odsp-urlresolver.api.md:13]: Seems like expected. Resolving this one...

File: packages/drivers/odsp-urlResolver/package.json (hunk 1 above)
Comment (chentong7) [reply packages/drivers/odsp-urlResolver/package.json]: I see. Adding type test file back

**Comment by chentong7** [comment]
> A few comments. Also, can we make the description clearer? I first thought it meant the whole package, but we're still removing it. After reviewing I understood it means the class that was moved.

Yes! We are just moving `FluidAppOdspUrlResolver` class to another dir so we delete package `fluidapp-odsp-urlresolver`. Renaming the PR name here

File: pnpm-lock.yaml [hunk 5]
```diff
@@ -45327,7 +45272,7 @@ packages:
     version: 0.1.0
     hasBin: true
     dependencies:
-      '@rushstack/node-core-library': 3.61.0
+      '@rushstack/node-core-library': 3.63.0
```
Comment (jatgarg) [inline pnpm-lock.yaml:45275]: Can you revert his change? I don't think this is required?

File: pnpm-lock.yaml [hunk 6]
```diff
@@ -20152,7 +20097,7 @@ packages:
       assert-never: 1.2.1
       globby: 13.2.2
       prettier: 3.0.3
-      typescript: 5.3.2
+      typescript: 5.3.3
```
Comment (jatgarg) [inline pnpm-lock.yaml:20100]: Can you revert his change? I don't think this is required?

File: packages/drivers/odsp-urlResolver/src/odspUrlResolver.ts
Comment (alexvy86) [inline packages/drivers/odsp-urlResolver/src/odspUrlResolver.ts:1]: What's the reason for the changes in this file? Moving to `Set`, and moving from `substring()` to `slice()`?

File: packages/drivers/odsp-urlResolver/package.json [hunk 7]
```diff
@@ -95,14 +95,17 @@
...
 		"@fluidframework/core-interfaces": "workspace:~",
+		"@fluidframework/core-utils": "workspace:~",
 		"@fluidframework/driver-definitions": "workspace:~",
 		"@fluidframework/odsp-driver": "workspace:~",
 		"@fluidframework/odsp-driver-definitions": "workspace:~"
 	},
 	"devDependencies": {
 		"@arethetypeswrong/cli": "^0.13.3",
 		"@fluid-tools/build-cli": "^0.29.0",
+		"@fluid-tools/fluidapp-odsp-urlresolver-previous": "npm:@fluid-tools/fluidapp-odsp-urlresolver@2.0.0-internal.8.0.0",
```
Comment (alexvy86) [inline packages/drivers/odsp-urlResolver/package.json:108]: @chentong7 can you create an ADO item to track removing this? It should only be necessary right now because of the move from the removed package to the new one, but it should be removed in the next version bump, once we have a published "previous" for odsp-urlresolver, right?

Comment (Josmithr) [reply packages/drivers/odsp-urlResolver/package.json:108]: Have we verified that this works? To my knowledge, the type-tests only have awareness of the current package name, and look for `<package-name>-previous`. So I don't think they will ever see this dependency (worth double checking, but I'm pretty sure this is the case). In which case I'm not sure we get much benefit from adding this dep and moving over the type-tests from the old package.

FWIW, I don't think maintaining API compatibility with the old package is even needed.

Comment (Josmithr) [reply packages/drivers/odsp-urlResolver/package.json:108]: To clarify, what I mean is: Adding this dependency and copying over the generated type-tests module from the old package will ensure that running typescript builds the file we copied over, but no new changes will be detected for that module, since the type-test infrastructure doesn't know it exists. So, I'm not sure keeping it adds much benefit.

File: packages/drivers/odsp-urlResolver/src/odspUrlResolver.ts
Comment (chentong7) [reply packages/drivers/odsp-urlResolver/src/odspUrlResolver.ts:1]: Eslint's error

File: pnpm-lock.yaml (hunk 6 above)
Comment (chentong7) [reply pnpm-lock.yaml:20100]: Sure! It came with rebase and re-built, is that any risk to upgrade the package?

File: pnpm-lock.yaml (hunk 5 above)
Comment (chentong7) [reply pnpm-lock.yaml:45275]: Sure

File: packages/drivers/odsp-urlResolver/package.json (hunk 7 above)
Comment (chentong7) [reply packages/drivers/odsp-urlResolver/package.json:108]: Yes. I did verify it by `git clean -dxf` and then rebase and rebuilt from root. We do still have this package previous in pnpm module after that. Then should I exclude the type test to this class?

Comment (chentong7) [reply packages/drivers/odsp-urlResolver/package.json:108]: Should I move FluidAppOdspUrlResolver class to existing file `urlResolver.ts` and add type test into this file?

Comment (chentong7) [reply packages/drivers/odsp-urlResolver/package.json]: Updated

File: packages/drivers/odsp-urlResolver/src/test/fluidAppUrlResolver.spec.ts [hunk 8]
```diff
@@ -3,7 +3,7 @@
  * Licensed under the MIT License.
  */
 
-import { strict as assert } from "assert";
+import { strict as assert } from "node:assert";
```
Comment (jatgarg) [inline packages/drivers/odsp-urlResolver/src/test/fluidAppUrlResolver.spec.ts:6]: Why node:assert? I think we just import from "assert" at other places.

**Review by jatgarg — APPROVED** [review]

File: packages/drivers/odsp-urlResolver/src/test/fluidAppUrlResolver.spec.ts (hunk 8 above)
Comment (alexvy86) [reply packages/drivers/odsp-urlResolver/src/test/fluidAppUrlResolver.spec.ts:6]: With our stricter eslint configurations the explicit specifier is required for node modules. The new package where this ended up must already have a stricter one that we had in the removed package.

**Review by alexvy86 — COMMENTED** [review]
One last thing here, let's add a changeset file to announce the removal of the package. Getting rid of any packages is worth calling out on the release notes so a changeset file makes sense here.

**Comment by chentong7** [comment]
> One last thing here, let's add a changeset file to announce the removal of the package. Getting rid of any packages is worth calling out on the release notes so a changeset file makes sense here.

Added

File: .changeset/ripe-cooks-yawn.md [hunk 9]
```diff
@@ -0,0 +1,8 @@
+---
+"@fluid-tools/fetch-tool": minor
+"@fluidframework/odsp-urlresolver": minor
+---
```
Comment (alexvy86) [inline .changeset/ripe-cooks-yawn.md:4]: I _think_ the suggestion below is valid... can't recall if it might cause issues during the release process (maybe if we try to generate changelogs in the individual affected packages, since this one will be gone; cc @tylerbutler @dannimad @jatgarg @CraigMacomber to see if one of you remembers how we should announce packages going away).

In any case, I don't think we need to call out fetch-tool or odsp-urlresolver as affected.

```suggestion
---
"@fluid-tools/fluidapp-odsp-urlresolver": major
---
```

Comment (chentong7) [reply .changeset/ripe-cooks-yawn.md:4]: Well the `pnpm changeset` doesn't provide @fluid-tools/fluidapp-odsp-urlresolver as option when choosing package. Although I think it does make sense to have announcement in your example way.
